EC-84: Revise Configuration, Customers, eContent, Login and Logging (#5)
ezsystems/commerce-documentation: docs/guide/login_and_registration/login_and_registration_cookbook.md

# Login and registration cookbook

eZ Commerce is able to check credentials not only specified by username and password,
but also using customer number (which could be taken from ERP system). 

## Services

### AuthenticationListener

The first part in the authentication process is `AuthenticationListener`.
This listener is able to read the posted values from the login form and create a `UserToken`.

The default Symfony listener is extended to collect more information (`customer_number`).

> In the standard Symfony login form only username and password are allowed.

Reviewer: ```suggestion
In the standard Symfony login form, only username and password are allowed.
```

The extended listener implements the same `AbstractAuthenticationListener` interface.

`services.xml`:

``` xml
<parameter key="security.authentication.listener.form.class">Silversolutions\Bundle\EshopBundle\Service\Security\CustomFormAuthenticationListener</parameter>
```

A different `UsernamePasswordToken` (using `TokenInterface`) is needed here.

> You have to add new attribute to the `customer_no` token. This `UsernamePasswordToken` is passed to next service in the authentication process.

Reviewer: ```suggestion
You must add a new attribute to the `customer_no` token. This `UsernamePasswordToken` is passed to the next service in the authentication process.
```

### UserProvider

`UserProvider` gets the created `UsernamePasswordToken` and fetches the correct User from the content model.

> The default fetching functionality is extended because in the Customer Center the User can be placed in different User Groups (or outside a group).

Reviewer: ```suggestion
The default fetching functionality is extended because, in the Customer Center, the User can be placed in different User Groups (or outside a group).
```

The provider uses `locationId` in the backend to determine private/b2b customers.

> It provides a `checkEzUser()` method which checks the Location and customer number of the given User.

Reviewer: ```suggestion
It provides a `checkEzUser()` method that checks the Location and customer number of the given User.
```

``` xml
<parameter key="ezpublish.security.user_provider.class">Silversolutions\Bundle\EshopBundle\Service\Security\UserProvider</parameter>
```

### AuthenticationProvider

> `AuthenticationProvider` retrieves and authenticates the User (using the `UsernamePasswordToken`).

Reviewer: ```suggestion
`AuthenticationProvider` retrieves and authenticates the User (with the `UsernamePasswordToken`).
```

> The methods `retrieveUser()` and `checkAuthentication()` try to load a User by username or email considering the credentials.

Reviewer: Is it "Depending on the credentials provided, ..."?

Afterwards, also the Location and possibly the customer number are checked by the `checkEzUser()` method from `UserProvider`.

``` xml
<parameter key="security.authentication.provider.dao.class">Silversolutions\Bundle\EshopBundle\Service\Security\AuthenticationProvider</parameter>
Security Controller
```

## Security Controller

The goal of the Security Controller's `loginAction()` is to display the login mask and authentication errors, if there are any.

``` php
/**
 * renders the login form
 *
 * @param \Symfony\Component\HttpFoundation\Request $request
 * @return Response
 */
public function loginAction(Request $request)
{
    $session = $request->getSession();
    if ($request->attributes->has(SecurityContextInterface::AUTHENTICATION_ERROR)) {
        $error = $request->attributes->get(SecurityContextInterface::AUTHENTICATION_ERROR);
    } else  {
        $error = $session->get(SecurityContextInterface::AUTHENTICATION_ERROR);
        $session->remove(SecurityContextInterface::AUTHENTICATION_ERROR);
    }
    return $this->render(
        'SilversolutionsEshopBundle:Security:login.html.twig',
        array(
            'error' => $error,
        )
    );
}
```

## Login form

> Login form is adjusted to use additional parameter `_customer_no`.

Reviewer: ```suggestion
Login form is adjusted to use the additional parameter `_customer_no`.
```

    The login forms in the template must post to the `{{ path( 'login_check' ) }}`

    This URL is not the URL of the security controller.
    The posted data is handled automatically by Symfony and passed to the Security context, where the authentication is done.

## Logging in a User from a controller

Sometimes a User must be logged in directly, without providing the username/password combination.

> An example use case is logging in via SSO:

Reviewer: ```suggestion
An example use case is logging in with SSO:
```
